### PR TITLE
[ NNSTREAMER ] Add nnstreamer filter for nntrainer

### PR DESCRIPTION
**Pull request by jijoongmoon**
In this PR,

 . NNStreamer filter subplugin is implemented.
 . Test Cases for subplugin

**Self evaluation:**
1. Build test:	 [X]Passed [ ]Failed [ ]Skipped
2. Run test:	 [X]Passed [ ]Failed [ ]Skipped

Signed-off-by: jijoong.moon <jijoong.moon@samsung.com>

### CODE REVIEW

File: nnstreamer/tensor_filter/tensor_filter_nntrainer.cc [hunk 1]
```diff
@@ -0,0 +1,369 @@
...
+#endif
+  return 0;
+}
+
+void NNTrainer::freeOutputTensor(void *data) {
+  if (data != nullptr) {
+    std::map<void *, nntrainer::Tensor *>::iterator it =
+      outputTensorMap.find(data);
+    if (it != outputTensorMap.end()) {
+      std::cout << "asdfasdfasdfasdf" << std::endl;
```
Comment (kparichay) [inline nnstreamer/tensor_filter/tensor_filter_nntrainer.cc:248]: remove?

File: nnstreamer/tensor_filter/tensor_filter_nntrainer.cc [hunk 2]
```diff
@@ -0,0 +1,369 @@
...
+      delete it->second;
```
Comment (zhoonit) [inline nnstreamer/tensor_filter/tensor_filter_nntrainer.cc:249]: I think std::shared_ptr owns the data, is this safe to free this?

File: nnstreamer/tensor_filter/tensor_filter_nntrainer.cc [hunk 3]
```diff
@@ -0,0 +1,369 @@
...
+  gst_tensors_info_copy(info, &outputTensorMeta);
+  return 0;
+}
+
+int NNTrainer::run(const GstTensorMemory *input, GstTensorMemory *output) {
+#if (DBG)
+  gint64 start_time = g_get_real_time();
+#endif
+  std::vector<nntrainer::Tensor *> output_tensors;
+  nntrainer::Tensor *out = new nntrainer::Tensor();
```
Comment (kparichay) [inline nnstreamer/tensor_filter/tensor_filter_nntrainer.cc:213]: this is over-written by the line 230. 
Memory leak?

File: nnstreamer/tensor_filter/tensor_filter_nntrainer.cc [hunk 4]
```diff
@@ -0,0 +1,369 @@
...
+    model->finalize();
+    delete out;
+    return -1;
+  }
+
+  *out = *o;
+
+  output[0].data = out->getData();
+
+  outputTensorMap.insert(std::make_pair(output[0].data, output_tensors[0]));
```
Comment (kparichay) [inline nnstreamer/tensor_filter/tensor_filter_nntrainer.cc:234]: IMPORTANT: as the data in the output tensor is a shared_ptr, storing just the data ptr wont suffice. The shared_ptr can be freed in the background
need to store the shared_ptr itself so as to stop it from freeing.

**Review by kparichay — APPROVED** [review]
no functionality issues
others can be dealt with later

**Review by zhoonit — APPROVED** [review]
LGTM

**Pull request merged**